src/reactapp/src/components/billingAddress/components/BillingAddressForm.jsx

```diff
@@ -33,100 +44,119 @@ function BillingAddressForm() {
   });
   const { selectedCountry, isBillingAddressTouched } = formikData;
 
+  const saveAddressAction = async () => {
+    await formSubmitHandler();
+
+    if (!isLoggedIn) {
+      return;
+    }
+
+    if (isNewAddress) {
+      const recentAddressList = LocalStorage.getMostlyRecentlyUsedAddressList();
+      const newAddressId = `new_address_${_keys(recentAddressList).length + 1}`;
+      LocalStorage.addAddressToMostRecentlyUsedList(billingValues);
+      setIsNewAddress(false);
+      setSelectedAddress(newAddressId);
+      LocalStorage.saveCustomerAddressInfo(newAddressId, isBillingSame, false);
+      reCalculateMostRecentAddressOptions();
+    }
+
+    if (isMostRecentAddress(selectedAddress)) {
+      LocalStorage.updateMostRecentlyAddedAddress(
+        selectedAddress,
+        billingValues
+      );
+      reCalculateMostRecentAddressOptions();
+    }
+  };
+
   if (viewMode) {
     return <></>;
   }
 
-  if (isBillingAddressSameAsShipping) {
-    return <BillingSameAsShippingCheckbox />;
-  }
-
   return (
     <>
-      <BillingSameAsShippingCheckbox />
-      <ORBox />
-
       <div className="py-2">
         <TextInput
           required
           label={__('Company')}
           name={fields.company}
+          formikData={formikData}
```

Comment:
Wondering why formikData is passed here, would you mind explaining me?

Reply:
yes, this is for performance reasons. Using `useFormikContext` inside every component was the old way. But this will re-render that component each time there is a change in formik state. For example, we are filling out the email address field => that changes formik states => that will re-render address, payment, shipping section etc. 

I revamped this using `formikData`. formikData only contains formik state related to the particular section. We are wrapping each section with `React.memo` and thus the individual sections will be rendered only when the corresponding `formikData` changes. Please check the `BillingAddress` and `BillingAddressMemorized` to understand this in detail.

Reply:
Component defines the billing address section
```
function BillingAddress() {
  const { values } = useFormikContext();
  const formSectionData = useFormikMemorizer(BILLING_ADDR_FORM);
  const selectedRegion = _get(values, regionField);
  const selectedCountry = _get(values, countryField);
  const isBillingSame = _get(values, billingSameAsShippingField);
  const { formSectionValues, isFormSectionTouched } = formSectionData;
  const billingOtherOptionSelected = _get(values, billingAddrOtherOptionField);

  const billingFormikData = useMemo(
    () => ({
      ...formSectionData,
      isBillingSame,
      selectedRegion,
      selectedCountry,
      billingOtherOptionSelected,
      billingValues: formSectionValues,
      isBillingAddressTouched: isFormSectionTouched,
    }),
    [
      isBillingSame,
      selectedRegion,
      formSectionData,
      selectedCountry,
      formSectionValues,
      isFormSectionTouched,
      billingOtherOptionSelected,
    ]
  );

  return <BillingAddressMemorized formikData={billingFormikData} />;
}
```
Here we are preparing the formikData and feeding the component `BillingAddressMemorized`. Now see the memorized component

```
const BillingAddressMemorized = React.memo(({ formikData }) => {
  const { isBillingSame } = formikData;

  return (
    <BillingAddressFormikProvider formikData={formikData}>
      {!isBillingSame && (
        <Card>
          <ToggleBox title={__('Billing Information')} show>
            <BillingAddressForm />
            <BillingAddressView />
          </ToggleBox>
        </Card>
      )}
    </BillingAddressFormikProvider>
  );
});
```
Basically in this way, all children component of billing address section will be re-rendered if there is a change in the `formikData`.

Finally we will be passing the formikData through the formik provider component. In this case it is `BillingAddressFormikProvider`. If you check in this component, you will see the passing context will hold this.

```
const context = { formikData, ...formikData }
```
This way, now all child component can access formikData through the `BillingAddressFormikProviderContext`.

Reply:
Also, in particular, for the form inputs, we were using another helper `useField` which will re-render the form inputs always. So I removed that logic from TextInput, Checkbox, SelectInput, RadioInput, etc, and serve them using `formikData`. This way, the inputs in another section won't be re-rendered unnecessarily. 

We can further improve the performance by creating memorized component for each form input component. But, that is "too much" work I feel. App seems to perform well now. So subtle level of performance improvements can be avoided at this stage.

Reply:
thanks for the detailed explanation!